### ISSUE TITLE
[Bug]: The first page of the PDF is displayed at the last in Mozilla browsers

### ISSUE THREAD

**Issue by harishfalco** [mozilla/pdf.js#19251]
### Attach (recommended) or Link to PDF file

Issue: In Firefox browsers, the first page of the PDF is rendered as the last page. This issue is happening for some PDF files on Firefox browsers on Windows and Mac. On iPhone , it works fine.

Code Used

            const pdfjsLib = window.pdfjsLib;
            pdfjsLib.GlobalWorkerOptions.workerSrc = `https://cdnjs.cloudflare.com/ajax/libs/pdf.js/${window.pdfjsLib.version}/pdf.worker.min.mjs`;

            const blobUrl = URL.createObjectURL(imgFile);

            const loadingTask = pdfjsLib.getDocument({
              url: blobUrl,
              disableAutoFetch: true, // Disable auto-fetching of additional data
              disableStream: true, // Disable streaming to prevent partial loading
              isEvalSupported: false, // Disable JavaScript execution
            });
            loadingTask.promise.then(function (pdf) {
              const numPages = pdf.numPages;
              for (let pageNum = 1; pageNum <= numPages; pageNum++) {
                pdf.getPage(pageNum).then(function (page) {
                  const viewport = page.getViewport({ scale: 1.0 });
                  const desiredWidth = fileContainer.offsetWidth;
                  const scale = desiredWidth / viewport.width;
                  const scaledViewport = page.getViewport({ scale: scale });
                  const canvas = document.createElement('canvas');
                  const context = canvas.getContext('2d');
                  const outputScale = window.devicePixelRatio || 1;

                  canvas.style.width = '100%';
                  canvas.style.height = '210%';
                  canvas.width = Math.floor(scaledViewport.width * outputScale);
                  canvas.height = Math.floor(scaledViewport.height * outputScale);

                  context.scale(outputScale, outputScale);

                  const renderContext = {
                    canvasContext: context,
                    viewport: scaledViewport,
                  };

                  page.render(renderContext).promise.then(() => {
                    pdfContainer.appendChild(canvas);
                    if (pageNum === numPages) {
                      URL.revokeObjectURL(blobUrl);
                    }
                  }).catch(function (error) {
                    sendErrorToSentry(error);
                    showPDFError();
                  })
                });
              }
            }).catch(function (error) {
              console.log('error in pdfjs', error);
              sendErrorToSentry(error);
              // showPDFError();
            });

Sample PDF FIle (removed)

The Page with SBI Logo is the first page. In firefox borwsers on windows and Mac , the first page will be rendered at the last

PDF.JS version used
https://cdnjs.cloudflare.com/ajax/libs/pdf.js/4.9.155/pdf.min.mjs
https://cdnjs.cloudflare.com/ajax/libs/pdf.js/4.9.155/pdf.worker.min.mjs

### Web browser and its version

Firefox (on Mac) and Version : 133.0.3

### Operating system and its version

MacOS , version : 13.2.1 (22D68)

### PDF.js version

4.9.155

### Is the bug present in the latest PDF.js version?

Yes

### Is a browser extension

Yes

### Steps to reproduce the problem

1. Open the attached PDF file using PDF.js in Firefox Browsers on Windows or Mac.
2. You can see that first page is displayed at the end Firefox

### What is the expected behavior?

1. The first page should be displayed at the begining

### What went wrong?

Not very sure on what went wrong.

**Comment by timvandermeij** [comment]
I have removed the PDF file from the issue description because there is PII (personally identifiable information) in there and I doubt it was intended to share that. Note that covering the text in PDF documents does _not_ actually remove the text; it just makes the text invisible, but text selection/extraction of the underlying text is still possible.

For what it's worth, I cannot reproduce the issue with Firefox on Linux using https://mozilla.github.io/pdf.js/web/viewer.html, but the bug report also only mentions Windows and Mac so that probably makes sense.

I would recommend, before we can reproduce this issue, to upload a PDF document that is not sensitive to avoid the PII issue and so we can hopefully use it for our test suite to prevent regressions.

**Comment by harishfalco** [comment]
Hey @timvandermeij. Unfortunately, the issue is not happening on any of the PDFs that I tried.
I wanted to confirm if this issue is restricted to that file only or others as well. What would be the way to find it? 

Can I share the file with your email only so you can investigate the root cause?

**Comment by nicolo-ribaudo** [comment]
On MacOS, you can use the "Redact" tool of the built-in PDF viewer yo actually remove the personal data from the PDF.

**Comment by timvandermeij** [comment]
> What would be the way to find it?

If the issue doesn't occur for any other PDF file it's indeed likely that there is something specific to this PDF file.

> Can I share the file with your email only so you can investigate the root cause?

I don't have access to Windows or Mac, and the problem doesn't occur on Linux, so I don't think I can reproduce this. The reason we normally ask for a public file that can be uploaded here is that users with such configurations can help out with debugging this. As Nicolò suggested, if no other file can be used it's sometimes an option to create a reduced PDF file without sensitive contents, but do make sure that the problem still occurs after sanitization because it can occur that there is some corruption in the original PDF file that is fixed by PDF editors when re-saving the file.

**Comment by Snuffleupagus** [comment]
> Code Used

Please note https://github.com/mozilla/pdf.js/blob/master/.github/CONTRIBUTING.md (emphasis mine):
> If you are developing a custom solution, first check the examples at https://github.com/mozilla/pdf.js#learning and search existing issues. If this does not help, **please prepare a short well-documented example that demonstrates the problem and make it accessible online** on your website, JS Bin, GitHub, etc. before opening a new issue or contacting us in the Matrix room -- **keep in mind that just code snippets won't help us troubleshoot the problem**.

Looking at your code it doesn't take the asynchronicity of the PDF.js methods into account, which could definitely lead to pages being placed out-of-order if you're just inserting them "blindly" into a container; something like the following would avoid that:
```javascript
const pdfjsLib = window.pdfjsLib;
pdfjsLib.GlobalWorkerOptions.workerSrc = `https://cdnjs.cloudflare.com/ajax/libs/pdf.js/${window.pdfjsLib.version}/pdf.worker.min.mjs`;

const blobUrl = URL.createObjectURL(imgFile);

const loadingTask = pdfjsLib.getDocument({
  url: blobUrl,
  disableAutoFetch: true, // Disable auto-fetching of additional data
  disableStream: true, // Disable streaming to prevent partial loading
  isEvalSupported: false, // Disable JavaScript execution
});

try {
  const pdf = await loadingTask.promise;
  const numPages = pdf.numPages;
  for (let pageNum = 1; pageNum <= numPages; pageNum++) {
    const page = await pdf.getPage(pageNum);
    const viewport = page.getViewport({ scale: 1.0 });
    const desiredWidth = fileContainer.offsetWidth;
    const scale = desiredWidth / viewport.width;
    const scaledViewport = page.getViewport({ scale: scale });
    const canvas = document.createElement('canvas');
    const context = canvas.getContext('2d');
    const outputScale = window.devicePixelRatio || 1;

    canvas.style.width = '100%';
    canvas.style.height = '210%';
    canvas.width = Math.floor(scaledViewport.width * outputScale);
    canvas.height = Math.floor(scaledViewport.height * outputScale);

    context.scale(outputScale, outputScale);

    const renderContext = {
      canvasContext: context,
      viewport: scaledViewport,
    };

    try {
      await page.render(renderContext).promise;
      pdfContainer.appendChild(canvas);

      if (pageNum === numPages) {
        URL.revokeObjectURL(blobUrl);
      }
    } catch(error) {
      sendErrorToSentry(error);
      showPDFError();
    }
  }
} catch(error) {
  console.log('error in pdfjs', error);
  sendErrorToSentry(error);
  // showPDFError();
}
```

**Comment by harishfalco** [comment]
@Snuffleupagus,

Thanks for suggesting a potential solution. I’ve ensured .then is used wherever an asynchronous operation is expected. However, when I tried your solution, it’s throwing an Invalid canvas size error when attempting to render the PDF on screen.

Here’s a sample PDF where the issue occurs:
[sample-pdf.pdf](https://github.com/user-attachments/files/18224867/sample-pdf.pdf)

@timvandermeij  - This problem is specifically happening on Firefox for both Windows and Mac. Interestingly, on Mac, the issue doesn’t occur when the same document is uploaded for the second time.

Could you please suggest any workarounds or ways to debug this further?

**Comment by harishfalco** [comment]
@Snuffleupagus - Can you please re-open this issue as this is major blocker for me to use PDF.js ? Let me know if anyother information is requried. Thanks

**Comment by harishfalco** [comment]
@Snuffleupagus , May I know why both the issues are closed? In the new issue that was raised , I have attached a valid PDF file in which the issue is replicable
